### PR TITLE
Add script for hard-deleting accounts

### PR DESCRIPTION
**Pull request by Will-Howard**
This adds a script for permanently deleting accounts, mainly intended to be used for GDPR deletion requests. The script does the following:
- Sets `deleted` to true on the user to run the callbacks associated with soft deleting a user. These are:
  - Unsubscribing (but not hard deleting) from the digest
  - Updating elasticsearch to remove references to the user
- [Permanently deletes](https://mailchimp.com/developer/marketing/api/list-merges/) them from the two mailchimp lists they might be on (the digest, and the "Forum" audience we use for the drip campaign)
- Revokes their authorization for the forum in Auth0, and then deletes the Auth0 user if they aren't authorized for anything else

There is an annoying feature in mailchimp where if you permanently delete a user, [they can't then sign back up via the api](https://stackoverflow.com/questions/52198510/mailchimp-resubscribe-a-deleted-member-causes-the-api-to-return-a-400-bad-reques), only via a form (including embedded forms). This means if a user deletes their account and then re-signs up they won't be added to the drip campaign or be able to sign up to the digest. I have just made sure that trying to sign up for the digest gives a sensible error, but haven't tried to fix it apart from that.

I am planning to update the account settings page to add the option for users to do this themselves (with a cooling off period).

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207705757470373) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/users/helpers.tsx [hunk 1]
```diff
@@ -551,14 +551,25 @@ export const isMod = (user: UsersProfile|DbUser): boolean => {
 // TODO: I (JP) think this should be configurable in the function parameters
 /** Warning! Only returns *auth0*-provided auth0 Ids. If a user has an ID that
  * we get from auth0 but is ultimately from google this function will throw. */
-export const getAuth0Id = (user: DbUser) => {
+export const getAuth0IdIfUsernamePassword = (user: DbUser) => {
```
Comment (Will-Howard) [inline packages/lesswrong/lib/collections/users/helpers.tsx:554]: This only allowed the "auth0" provider so I added a version which allows any provider. I have made a ticket to come back and combine them as I suspect the original restriction to "auth0" isn't necessary in most cases

File: packages/lesswrong/server/users/permanentDeletion.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: `API_KEY ${mailchimpAPIKey}`,
+        },
+      }
+    );
+
+    if (!response.ok) {
+      const json = await response.json();
+      if (json.status !== 404) { // 404 just means they weren't subscribed to begin with
```
Comment (jpaddison3) [inline packages/lesswrong/server/users/permanentDeletion.ts:48]: k I guess that's a choice they could make

File: packages/lesswrong/server/users/permanentDeletion.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  await updateMutator({
+    collection: Users,
+    documentId: user._id,
+    set: { deleted: true },
+    validate: false,
+    context: adminContext,
+    currentUser: adminContext.currentUser,
+  })
+  // Wait until async callbacks finish. This is overcautious, as there should be no need for the callbacks to refetch the user object
+  await new Promise(resolve => setTimeout(resolve, 5000));
```
Comment (jpaddison3) [inline packages/lesswrong/server/users/permanentDeletion.ts:81]: In the shell script you can use a "wait for no pending callbacks". Obviously won't work when you go to the server.

File: packages/lesswrong/server/users/permanentDeletion.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+  // Remove from mailchimp lists
+  const mailchimpForumDigestListId = mailchimpForumDigestListIdSetting.get();
+  const mailchimpEAForumListId = mailchimpEAForumListIdSetting.get();
+
+  const listIdsToDeleteFrom = [mailchimpEAForumListId, mailchimpForumDigestListId].filter(v => v) as string[]
```
Comment (jpaddison3) [inline packages/lesswrong/server/users/permanentDeletion.ts:87]: ~The fuck is that filter?~ Fake edit: I am dumb and glad I asked my LLM.

**Review by jpaddison3 — APPROVED** [review]
LGTM

File: packages/lesswrong/server/callbacks/userCallbacks.tsx [hunk 5]
```diff
@@ -284,68 +284,82 @@ getCollectionHooks("Users").editSync.add(async function usersEditCheckEmail (mod
...
-getCollectionHooks("Users").editAsync.add(async function subscribeToForumDigest (newUser: DbUser, oldUser: DbUser) {
+  const handleErrorCase = (errorMessage: string) => {
+    // If the user is deactivating their account, allow the update to continue. Otherwise,
+    // the user is explicitly trying to update their subscription, so throw and block the update
+    const err = new Error(errorMessage)
+    captureException(err)
+    if (!deactivatedAccount) {
+      throw err
+    }
+    data.subscribedToDigest = false
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/callbacks/userCallbacks.tsx:309]: It's so the field on the user remains `false` we failed to sign them up for the digest, so it should now be a true reflection of whether they are signed up. Previously the two could get out of sync

File: packages/lesswrong/server/users/permanentDeletion.ts (hunk 3 above)
Comment (Will-Howard) [reply packages/lesswrong/server/users/permanentDeletion.ts:81]: Yep the fact that this needs to run on the live server means we can't use that, and I think it's better to handle the two cases in the same way

File: packages/lesswrong/server/authentication/auth0.ts [hunk 6]
```diff
@@ -137,22 +156,122 @@ class Auth0Client extends IAuth0BackendClient {
...
+    return false;
+  }
+
+  await auth0Client.deleteUser(auth0UserId)
+  return true;
+}
+
 // TODO: Probably good to fix this, IM(JP)O. It works because we only use it in
 // a context where we're guaranteed to have an email/password user.
 /** Warning! Only returns profiles of users who do not use OAuth */
```
[the remark this answers is not in the document]
Comment (Will-Howard) [reply packages/lesswrong/server/authentication/auth0.ts:267]: Oh this is still correct, pending [this](https://app.asana.com/0/1124817936128791/1207705410755081/f)

**Pull request merged**